### ISSUE TITLE
NF-UNSW-Dataset to Spiking

### ISSUE THREAD

**Issue by onlysubgroup** [neuromorphs/tonic#196]
Hi,

so I want to add the NSL-KDD Dataset(https://www.kaggle.com/datasets/hassan06/nslkdd) to tonic and make it spiking.
I found hdf5 formatted and preprocessed files of the dataset in https://github.com/kyuyeonpooh/nsl-kdd-autoencoder/tree/master/hdf5  .
Any advice besides, what's in https://tonic.readthedocs.io/en/latest/about/contribute.html ?

Greetings & Thanks in advance,
onlysubgroup

**Comment by biphasic** [comment]
Hey @onlysubgroup, that's really interesting! So far I've only worked with neuromorphic sensor datasets, so I'm curious where this leads to. The fact that hdf5 files are available is already great. Let's hope Github won't block too many downloads. 
I guess the crucial question is how to make it spiking. Have you thought about this yet? Having had a quick look at NSL-KDD it seems as if there are some 40 features per data point. That seems like a lot of dimensions per spike! Also, are you planning to make use of timestamps?

**Comment by biphasic** [comment]
You will need some kind of sequential / temporal nature I'd say. Are you planning to use a specific model (like a spiking neural network)?

**Comment by onlysubgroup** [comment]
Hi,

current State of the Art datasets like "NetFlow V2 Datasets" ("https://staff.itee.uq.edu.au/marius/NIDS_datasets/#RA5") use NetFlows to identify intrusions.
But still each flow is considered on it's own.
I will try to rate-code the proposed features and see if it makes sense to use SNNs in this context.

Cheers,
onlysubgroup

**Comment by onlysubgroup** [comment]
First tests with tonic and norse:
https://colab.research.google.com/drive/140OvbVrlEN4Cy_uK18EBk15kb80g1060?usp=sharing

**Comment by biphasic** [comment]
I don't have NF-UNSW_train.h5, where did you get it from? Also, is this a single sample? What are the features encoded?

**Comment by onlysubgroup** [comment]
Ah sry, should have also shared to preprocessing, my bad. But it's kind of messy still:
https://colab.research.google.com/drive/1KHxPQ21aRPAFfXT7ZgXxP9U_AsbCpDgi?usp=sharing

**Comment by biphasic** [comment]
Looking at the different features, I think one option would be to combine some of them. 

How about you encode IPV4_SRC_ADDR and L4_SRC_PORT as source neuron location? Basically get a list of all combinations of those columns and that's the amount of neurons you encode it with. 
Same for IPV4_DST_ADDR and L4_DST_PORT for the target neurons.

Then you make them fire at certain times, maybe OUT_BYTES, OUT_PKTS and FLOW_DURATION_MILLISECONDS could be combined to a rate code of spikes? Plus connections in the other direction for IN_BYTES, IN_PKTS

Not sure what to do about PROTOCOL, L7_PROTO, TCP_FLAGS.

I assume you want to learn to predict the label/attack.

**Comment by onlysubgroup** [comment]
Thanks for the reply and the ideas.
Yes, I want to predict the label/attack.
I will try it out and keep you up to date.

**Comment by onlysubgroup** [comment]
> Looking at the different features, I think one option would be to combine some of them.
> 
> How about you encode IPV4_SRC_ADDR and L4_SRC_PORT as source neuron location? Basically get a list of all combinations of those columns and that's the amount of neurons you encode it with. Same for IPV4_DST_ADDR and L4_DST_PORT for the target neurons.
> 
> Then you make them fire at certain times, maybe OUT_BYTES, OUT_PKTS and FLOW_DURATION_MILLISECONDS could be combined to a rate code of spikes? Plus connections in the other direction for IN_BYTES, IN_PKTS
> 
> Not sure what to do about PROTOCOL, L7_PROTO, TCP_FLAGS.
> 
> I assume you want to learn to predict the label/attack.

Hey,

I looked up the Netflow Paper again (https://arxiv.org/pdf/2011.09144.pdf), they advice not to encode src/dst ip/port because it can lead to overfitting. I will just rate code the other features and see if I can get similar results, like the authors of the aforementioned paper.

Cheers,
onlysubgroup

**Comment by biphasic** [comment]
closing this for now. Please get in touch on the [Discord channel](https://discord.gg/V6FHBZURkg) if you have further questions!
